From 670c5363545f17a65066e6f0f25ba9dfb4e243f7 Mon Sep 17 00:00:00 2001
From: Travis Skare <skare@chromium.org>
Date: Thu, 23 Jun 2022 16:49:04 +0000
Subject: [PATCH] [QrCode] Skip WRITE_EXTERNAL_STORAGE permission request on
 Android T+

Tested QR code as well as long-press -> download image functionalty on webpages as this affects FileAccessPermissionHelper.

Bug: 1307347
Change-Id: Ifb3fccee4acce119263c9d351999d8042756a734
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3645848
Commit-Queue: Travis Skare <skare@chromium.org>
Reviewed-by: Min Qin <qinmin@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1017209}
---
 .../browser/download/FileAccessPermissionHelper.java     | 4 +++-
 .../chrome/browser/share/SaveBitmapDelegate.java         | 7 +++++--
 .../share/qrcode/share_tab/QrCodeShareMediator.java      | 9 +++++++++
 3 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/chrome/android/java/src/org/chromium/chrome/browser/download/FileAccessPermissionHelper.java b/chrome/android/java/src/org/chromium/chrome/browser/download/FileAccessPermissionHelper.java
index 2aedd0c6c69c7..0f2b878c46148 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/download/FileAccessPermissionHelper.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/download/FileAccessPermissionHelper.java
@@ -11,6 +11,7 @@ import android.util.Pair;
 
 import androidx.annotation.NonNull;
 
+import org.chromium.base.BuildInfo;
 import org.chromium.base.Callback;
 import org.chromium.base.Consumer;
 import org.chromium.components.permissions.AndroidPermissionRequester;
@@ -51,7 +52,8 @@ public class FileAccessPermissionHelper {
 
     static void requestFileAccessPermissionHelper(
             @NonNull WindowAndroid windowAndroid, final Callback<Pair<Boolean, String>> callback) {
-        if (windowAndroid.hasPermission(permission.WRITE_EXTERNAL_STORAGE)) {
+        if (BuildInfo.isAtLeastT()
+                || windowAndroid.hasPermission(permission.WRITE_EXTERNAL_STORAGE)) {
             callback.onResult(Pair.create(true, null));
             return;
         }
diff --git a/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/SaveBitmapDelegate.java b/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/SaveBitmapDelegate.java
index be7fde3e1184d..e7ff266387169 100644
--- a/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/SaveBitmapDelegate.java
+++ b/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/SaveBitmapDelegate.java
@@ -17,6 +17,7 @@ import android.provider.Settings;
 import androidx.annotation.VisibleForTesting;
 import androidx.appcompat.app.AlertDialog;
 
+import org.chromium.base.BuildInfo;
 import org.chromium.chrome.R;
 import org.chromium.chrome.browser.download.FileAccessPermissionHelper;
 import org.chromium.ui.base.WindowAndroid;
@@ -60,8 +61,10 @@ public class SaveBitmapDelegate {
             return;
         }
 
-        if (!mWindowAndroid.hasPermission(permission.WRITE_EXTERNAL_STORAGE)
-                && !mWindowAndroid.canRequestPermission(permission.WRITE_EXTERNAL_STORAGE)) {
+        boolean needPermissionRequestButBlocked = !BuildInfo.isAtLeastT()
+                && !mWindowAndroid.hasPermission(permission.WRITE_EXTERNAL_STORAGE)
+                && !mWindowAndroid.canRequestPermission(permission.WRITE_EXTERNAL_STORAGE);
+        if (needPermissionRequestButBlocked) {
             AlertDialog.Builder builder =
                     new AlertDialog.Builder(mContext, R.style.ThemeOverlay_BrowserUI_AlertDialog);
             builder.setMessage(R.string.sharing_hub_storage_disabled_text)
diff --git a/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/qrcode/share_tab/QrCodeShareMediator.java b/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/qrcode/share_tab/QrCodeShareMediator.java
index 00c9076b1f156..3ca4fc5517fb4 100644
--- a/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/qrcode/share_tab/QrCodeShareMediator.java
+++ b/chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/qrcode/share_tab/QrCodeShareMediator.java
@@ -17,6 +17,7 @@ import android.text.TextUtils;
 import android.text.TextUtils.TruncateAt;
 import android.view.View;
 
+import org.chromium.base.BuildInfo;
 import org.chromium.base.metrics.RecordUserAction;
 import org.chromium.chrome.R;
 import org.chromium.chrome.browser.download.FileAccessPermissionHelper;
@@ -119,8 +120,15 @@ class QrCodeShareMediator {
         }
     }
 
+    /** Returns whether we need to explicitly request a storage permission. */
+    private Boolean requiresAdditionalStoragePermission() {
+        return !BuildInfo.isAtLeastT();
+    }
+
     /** Returns whether the user has granted storage permissions. */
     private Boolean hasStoragePermission() {
+        // Not needed for newer SDKs; treat as granted implicitly.
+        if (!requiresAdditionalStoragePermission()) return true;
         return mContext.checkPermission(
                        permission.WRITE_EXTERNAL_STORAGE, Process.myPid(), Process.myUid())
                 == PackageManager.PERMISSION_GRANTED;
@@ -128,6 +136,7 @@ class QrCodeShareMediator {
 
     /** Returns whether the user can be prompted for storage permissions. */
     private Boolean canPromptForPermission() {
+        if (!requiresAdditionalStoragePermission()) return false;
         if (mWindowAndroid == null) return false;
         return mWindowAndroid.canRequestPermission(permission.WRITE_EXTERNAL_STORAGE);
     }
